### PR TITLE
Replace "frozen insert" with "insert & freeze" across GPDB

### PR DESCRIPTION
**Pull request by huansong**
Similar to #15127, now apply the same approach for all the "frozen insert" logic in GPDB, for two purposes:
1. Solve the remaining inconsistency hazard. For example, the bitmap LOV could still see the same problem (a repro shown below).
2. Align core function interface with upstream, including things like `heap_prepare_insert` and `toast_save_datum`. With that we also removed GPDB-specific function such as `frozen_heap_insert` and `XLogInsert_OverrideXid`.

Please review each commit individually.

A repro for the inconsistency in bitmap LOV table:
```sql
-- on session 1, do
create extension pageinspect;
drop table if exists tab_fi;
create table tab_fi(a int, b int) with (appendoptimized=true) distributed replicated;
create index tab_fi_idx on tab_fi using bitmap(b);
insert into tab_fi values(1, 1); 
select gp_segment_id, pg_switch_wal() is not null from gp_dist_random('gp_id') where gp_segment_id = 0;

-- launch GDB and attach to one of the QE process, say seg0 here.
-- breakpoint on frozen_heap_insert(), continue

-- on session 1, do
insert into tab_fi values(2, 2); 

-- on GDB session, step through frozen_heap_insert()

-- on another session 2, do
select gp_segment_id, pg_switch_wal() is not null from gp_dist_random('gp_id') where gp_segment_id = 0;
select gp_inject_fault('qe_exec_finished', 'panic', '', '', '', 1, -1, 0, dbid) from gp_segment_configuration where role = 'p' and content = 0;

-- exit the GDB session

-- connect to seg0 in utility mode, do:
set enable_indexscan = on;
set enable_seqscan = off;
select c.relname from bm_metap('tab_fi_idx') b join pg_class c on b.auxrelid = c.oid \gset
select b from pg_bitmapindex.:relname;
set enable_indexscan = off;
set enable_seqscan = on;
select b from pg_bitmapindex.:relname;

-- the two SELECTs above would show different results: the second one has one more row
```

dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/frozen-insert

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by soumyadeep2007 — COMMENTED** [review]
Some context for the last commit that might be helpful to be placed in the commit message:

```
The only call graph where a frozen insert to toast was necessary was (i.e.
toast_insert_or_update() called with isFrozen=true):

frozen_heap_insert() -> heap_insert() -> heap_prepare_insert() -> toast_insert_or_update()

gp_fastsequence, bitmap indexes and ao(cs)seg tables used this interface but
none of them have a toast table right? So, one may question why this change to
toast (and all the changes to toast functions throughout) was ever done.

Digging into history shows that this change was added to support an edge case
with "error tables" for external tables, which have been removed from GPDB since
3fd4bd9c6634d90f23099bdce75a910ba2813fad.

Eg of old syntax:
copy foo from 'foo.csv' csv LOG ERRORS INTO errtable SEGMENT REJECT LIMIT 10 ROWS;

So, all in all we can get rid of the freeze related changes to toast code.
```

File: src/test/isolation2/sql/frozen_insert_crash.sql [hunk 1]
```diff
@@ -98,6 +99,81 @@
...
+1: create index tab_fi_idx on tab_fi using bitmap(b);
+1: insert into tab_fi values(1, 1);
+1: select gp_segment_id, pg_switch_wal() is not null from gp_dist_random('gp_id') where gp_segment_id = 0;
+
+-- case 2: suspend and flush WAL after freezing the tuple
+1: select gp_inject_fault('insert_bmlov_after_freeze', 'suspend', ''/*DDL*/, ''/*DB*/, ''/*table*/, 1/*start occur*/, 1/*end occur*/, 0/*extra_arg*/, dbid) from gp_segment_configuration where role = 'p' and content = 0;
+2>: insert into tab_fi values(2, 2);
+1: select gp_wait_until_triggered_fault('insert_bmlov_after_freeze', 1, dbid) from gp_segment_configuration where role = 'p' and content = 0;
+1: select gp_segment_id, pg_switch_wal() is not null from gp_dist_random('gp_id') where gp_segment_id = 0;
+1: select gp_inject_fault('qe_exec_finished', 'panic', ''/*DDL*/, ''/*DB*/, ''/*table*/, 1/*start occur*/, -1/*end occur*/, 0/*extra_arg*/, dbid) from gp_segment_configuration where role = 'p' and content = 0;
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/test/isolation2/sql/frozen_insert_crash.sql]: Good catch! Done.

File: src/test/isolation2/sql/frozen_insert_crash.sql [hunk 2]
```diff
@@ -98,6 +99,81 @@
 
 1: drop table tab_fi;
 
+
+-- Same set of tests for bitmap LOV insert.
+-- Less comments about the test procedure here because it's
+-- almost exactly the same as the gp_fastsequence case.
+-- Please refer to the gp_fastsequence case for details about
+-- what we're doing here.
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/test/isolation2/sql/frozen_insert_crash.sql]: Makes sense.

File: src/backend/access/heap/heapam.c [hunk 3]
```diff
@@ -6676,9 +6676,12 @@ heap_freeze_tuple_wal_logged(Relation rel, HeapTuple tup)
 	heap_execute_freeze_tuple(htup, &frozen);
 
 	/* WAL logging */
-	recptr = log_heap_freeze(rel, buffer, InvalidTransactionId /* cutoff_xid */,
-								&frozen, 1 /*ntuples*/);
-	PageSetLSN(page, recptr);
+	if (RelationNeedsWAL(rel))
+	{
+		recptr = log_heap_freeze(rel, buffer, InvalidTransactionId /* cutoff_xid */,
```
Comment (ashwinstar) [inline src/backend/access/heap/heapam.c:6674]: can define `XLogRecPtr      	recptr;` then within this scope

File: src/backend/access/aocs/aocssegfiles.c [hunk 4]
```diff
@@ -96,7 +96,8 @@ InsertInitialAOCSFileSegInfo(Relation prel, int32 segno, int32 nvp, Oid segrelid
 
 	segtup = heap_form_tuple(RelationGetDescr(segrel), values, nulls);
 
-	CatalogTupleInsertFrozen(segrel, segtup);
+	CatalogTupleInsert(segrel, segtup);
```
Comment (ashwinstar) [inline src/backend/access/aocs/aocssegfiles.c:99]: Did we try concurrency scenario of having two transactions performing insert two AO table before first transaction freezes the tuple - can second transaction insert and pick same segfile for inserts? (will need fault injector between these two steps to pause one transactions execution and then see the effect)

Comment (huansong) [reply src/backend/access/aocs/aocssegfiles.c:99]: The second transaction will block at waiting to lock the AO table `choose_segno_internal->LockRelationForExtension` until the first transaction has finished choosing segno. But great point about a potential issue. Added a test for that.

File: src/test/isolation2/sql/frozen_insert_crash.sql [hunk 5]
```diff
@@ -98,6 +99,117 @@
...
+
+-- Test for aoseg: suspend the insert into aoseg table before we mark the row frozen.
+-- Another session should still be able to choose a different segno.
+1: create table tab_aoseg(a int) using ao_row;
+1: select gp_inject_fault('insert_aoseg_before_freeze', 'suspend', dbid) from gp_segment_configuration where role = 'p' and content = 0;
+1: begin;
+1>: insert into tab_aoseg select * from generate_series(1,10);
+2: begin;
+2>: insert into tab_aoseg select * from generate_series(1,10);
+3: select gp_inject_fault('insert_aoseg_before_freeze', 'reset', dbid) from gp_segment_configuration where role = 'p' and content = 0;
```
Comment (ashwinstar) [inline src/test/isolation2/sql/frozen_insert_crash.sql:206]: test is not checking if the suspend fault was hit or not, means leaving it to chance that specific scenario is really being exercised or not. Should wait for fault to hit before really starting the second insert stmt.

**Review by ashwinstar — APPROVED** [review]
:+1:

File: src/test/isolation2/sql/frozen_insert_crash.sql (hunk 5 above)
Comment (huansong) [reply src/test/isolation2/sql/frozen_insert_crash.sql:206]: Good catch! Fixed.

**Pull request merged**